sympy/matrices/random.py

```diff
@@ -0,0 +1,2343 @@
+                     length - int(length / 2))
+    return lwr.T * upr
+
+
+def invertible(dim,
+               scalars=_elementary_scalars,
+               units=_elementary_units,
+               length=None):
+    r"""an invertible matrix n x n
+
+    Explanation
+    ===========
+
+    Same as :func:`square` but with full rank, i.e. **rank=dim**.
+
+    Parameters
+    ==========
+    dim : integer
+        dimension of matrix
+    scalars : tuple or list of symbols (optional)
+        default values for random choosen scalar of
+        elementary matrices to build the matrix
+    units : tuple or list of symbols (optional)
+        default values for random choosen scalar of
+        elementary matrices to build the matrix
+        which are assumed to have a multiplicative inverse
+    length : integer (optional with default 2 * **dim**)
+        number of invertible matrices
+        to build the resulting matrix.
+
+    Examples
+    ========
+    >>> from sympy import symbols
+    >>> from sympy.core.random import seed
+    >>> from sympy.matrices.random import invertible
+    >>> seed(1)
+
+    >>> invertible(3)
+    Matrix([
+    [-1, -1, -1],
+    [-2, -1, -1],
+    [ 0,  0,  1]])
+
+    >>> phi = symbols('phi')
+    >>> m = invertible(3, scalars=(1, -1, phi, -phi))
+    >>> m
+    Matrix([
+    [1, 0,     -phi],
+    [1, 1, -phi - 1],
+    [0, 0,        1]])
+    >>> m.det()
+    1
+    >>> m.inv()
+    Matrix([
+    [ 1, 0, phi],
+    [-1, 1,   1],
+    [ 0, 0,   1]])
+
+    """
+    return square(dim, None, scalars, units, length)
+
+
+def singular(dim,
+             rank=None,
+             scalars=_elementary_scalars,
+             units=_elementary_units,
+             length=None):
+    r"""a singular square matrix n x n with a given rank.
+
+    Explanation
+    ===========
+
+    Same as :func:`square` but with rank defaulting to $dim-1$.
+
+    Examples
+    ========
+    >>> from sympy.core.random import seed
+    >>> from sympy.matrices.random import singular
+    >>> seed(1)
+
+    >>> m = singular(3, 2)
+    >>> m
+    Matrix([
+    [-1, 1, -1],
+    [-2, 2, -1],
+    [ 0, 0,  1]])
+    >>> m.rank()
+    2
+
+    Parameters
+    ==========
+    dim : integer
+        dimension of matrix
+    rank : integer (optional with default **dim**-1)
+        rank of matrix
+    scalars : tuple or list of symbols (optional)
+        default values for random choosen scalars of
+        elementary matrices to build the matrix
+    units : tuple or list of symbols (optional)
+        default values for random choosen scalar of
+        elementary matrices to build the matrix
+        which are assumed to have a multiplicative inverse
+    length : integer (optional with default 2 * **dim**)
+        number of invertible matrices to build the resulting matrix.
+
+    See Also
+    ========
+    invertible
+    triangular
+
+    """
+    rank = dim - 1 if rank is None else rank
+    return square(dim, rank, scalars, units, length)
```

Comment:
Maybe an error should be raised here if `rank == dim`

Reply:
Good point.